ci/src/cI_docker.ml

```diff
@@ -0,0 +1,143 @@
+open! Astring
+open Lwt.Infix
+open CI_utils.Infix
+
+module DK = CI_utils.DK
+
+module Image = struct
+  type t = {
+    id : string;
+  }
+
+  let v id = { id }
+
+  let id t = t.id
+
+  let pp f t = Fmt.string f t.id
+
+  let pp_short f t = Fmt.string f (String.with_range ~len:10 t.id)
+end
+
+(* Check that [path] is a valid path in [base] and contains no symlink components,
+   or "..". *)
+let validate_path ~log ~base path =
+  let rec aux base = function
+    | [] -> Ok ()
+    | "" :: _ -> Error "Empty path component"   (* Could be an absolute path *)
+    | ".." :: _ -> Error "'..' in Dockerfile path!"
+    | x :: xs ->
+      let subpath = Filename.concat base x in
+      match Unix.lstat subpath with
+      | {Unix.st_kind = Unix.S_REG | Unix.S_DIR; _} -> aux subpath xs
+      | _ -> Error (Fmt.strf "Not a regular file or directory: %S" x)
+      | exception Unix.Unix_error(Unix.ENOENT, _, _) -> Error "Dockerfile does not exist"
+      | exception ex -> Error (Fmt.strf "Bad Dockerfile: %a" CI_utils.pp_exn ex)
+  in
+  match aux base (String.cuts ~sep:"/" path) with
+  | Ok () -> ()
+  | Error msg ->
+    CI_live_log.log log "%s [%S]" msg path;
+    CI_utils.failf "%s" msg
+
+module Builder = struct
+
+  module Key = struct
+    type t = {
+      src : CI_git.commit;
+      from : Image.t option;
+    }
+  end
+
+  type t = {
+    label : string;
+    dockerfile : string;
+    timeout : float;
+    pool : CI_monitored_pool.t;
+  }
+
+  type context = CI_s.job_id
+
+  type value = Image.t
+
+  let name t =
+    Fmt.strf "docker build -f %S" t.dockerfile
+
+  let title _t _key = "Docker build"
+
+  let label t {Key.from; src} =
+    match from with
+    | None -> Fmt.strf "Build %s in %a" t.label CI_git.Commit.pp_short src
+    | Some base -> Fmt.strf "Build %s in %a from %a" t.label CI_git.Commit.pp_short src Image.pp_short base
+
+  let load _t tree _key =
+    DK.Tree.read_file tree CI_cache.Path.value >>*= fun data ->
+    Lwt.return (Image.v (String.trim (Cstruct.to_string data)))
+
+  let branch_safe_char = function
+    | ':' -> '-'
+    | x -> x
+
+  let branch t {Key.src; from} =
+    let from =
+      match from with
+      | None -> ""
+      | Some from -> "-from-" ^ String.map branch_safe_char (Image.id from)
+    in
+    Printf.sprintf "docker-build-%s-of-%s%s" t.label (CI_git.hash src) from
+
+  let rewrite_from ~log ?from path =
+    match from with
+    | None -> Lwt.return ()
+    | Some base ->
+      Lwt_io.with_file ~mode:Lwt_io.input path (fun ch -> Lwt_io.read ch) >>= fun contents ->
+      match String.cut ~sep:"\n" contents with
```

Comment:
Do we need to filter for `String.is_prefix ~affix:"#"` here as well, to account for comments before the `FROM` ?